infrastructure/src/main/java/org/corfudb/infrastructure/management/failuredetector/DecisionMakerAgent.java

```diff
@@ -0,0 +1,78 @@
+package org.corfudb.infrastructure.management.failuredetector;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.management.ClusterAdvisor;
+import org.corfudb.protocols.wireprotocol.ClusterState;
+import org.corfudb.protocols.wireprotocol.NodeState;
+import org.corfudb.protocols.wireprotocol.failuredetector.FileSystemStats;
+import org.corfudb.protocols.wireprotocol.failuredetector.FileSystemStats.PartitionAttributeStats;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank.NodeRankByPartitionAttributes;
+
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.TreeSet;
+
+@AllArgsConstructor
+@Slf4j
+public class DecisionMakerAgent {
+    private final ClusterState clusterState;
+    private final ClusterAdvisor clusterAdvisor;
+
+    /**
+     * Find a node that is a decision maker according to the current cluster state
+     *
+     * @return a decision maker
+     */
+    public Optional<String> findDecisionMaker() {
+        if (!clusterState.getLocalNode().isPresent()) {
+            return Optional.empty();
+        }
+
+        Optional<String> partitionDm = findPartitionAttributesDecisionMaker()
+                .map(NodeRankByPartitionAttributes::getEndpoint);
+
+        Optional<String> clusterDm = clusterAdvisor.findDecisionMaker(clusterState)
+                .map(NodeRank::getEndpoint);
+
+        //Only completely healthy nodes can update cluster layout
+        if (clusterDm.equals(partitionDm)) {
+            return clusterDm;
+        }
```

Comment:
Can this condition be false on all nodes in the cluster at once?  `clusterAdvisor.findDecisionMaker` will provide the node that has the largest number of active connections (with the lexicographically smallest name), but what happens if the partition on this node becomes read-only? Does that mean the failure detector will not be able to mark that node as unresponsive until another failure occurs?

Reply:
Correct. We can't do anything and a node can't make any changes in the cluster if the node itself is not completely healthy.
Our cluster of 3 nodes can survive only one failure. What you just described contains 2 failures in the cluster, we can't do anything in that situation, so we skip the current iteration of failure detection and hope that next time there will be only one failure in the cluster (in case of 3 nodes)

Reply:
As discussed offline, there could be a case when a node has the highest number of successful connections and it's read-only, in that case, the decision-maker on the other node must take into consideration that fact and exclude unhealthy nodes from the list, to allow other nodes to be a decision-maker